drivers/xen/gnttab.c

```diff
@@ -295,44 +299,67 @@ const char *gnttabop_error(int16_t status)
 	}
 }
 
+/* Picked from Linux implementation */
+#define LEGACY_MAX_GNT_FRAMES_SUPPORTED		4
+static unsigned long gnttab_get_max_frames(void)
+{
+	int ret;
+	struct gnttab_query_size q = {
+		.dom = DOMID_SELF,
+	};
+
+	ret = HYPERVISOR_grant_table_op(GNTTABOP_query_size, &q, 1);
+	if ((ret < 0) || (q.status != GNTST_okay)) {
+		return LEGACY_MAX_GNT_FRAMES_SUPPORTED;
+	}
+
+	return q.max_nr_frames;
+}
+
 static int gnttab_init(void)
 {
 	grant_ref_t gref;
 	struct xen_add_to_physmap xatp;
-	struct gnttab_setup_table setup;
-	xen_pfn_t frames[NR_GRANT_FRAMES];
 	int rc = 0, i;
+	unsigned long xen_max_grant_frames;
+	uintptr_t gnttab_base = DT_REG_ADDR_BY_IDX(DT_INST(0, xen_xen), 0);
+	mm_reg_t gnttab_reg;
+
```

Comment:
As far as I know, we can use only Grant Table Version 1 in this driver. I recommend checking that Zephyr OS is running with Grant Table Version 1 or explicitly requesting this version during initialization, as is done in the Linux Kernel driver: [Linux Kernel Grant Table Driver](https://elixir.bootlin.com/linux/v6.12.4/source/drivers/xen/grant-table.c#L1627).

I understand that currently, we have Version 1 hardcoded during domain creation. However, this behavior may change in the future. Therefore, it is better to notify the user about any version mismatch by checking the version or by always explicitly requesting Version 1 during driver initialization.

Reply:
Zephyr is currently supports only running on Arm64 and according to xenbits XSAs:

> On Arm, grant table v2 use is explicitly unsupported.

Probably more correct statement would be "security supported", because IIRC it can be enabled with some extra steps, but may lead to crashes. Thus, I see no urgent need to do this, especially in scope of this PR.

Reply:
This recommendation is primarily for future-proofing rather than addressing the current situation. For example, Xen might introduce a new version, such as Version 3, which could become the default. Alternatively, Version 2 might be fixed and set as the default for ARM. There's also a possibility that this driver might be extended to support x86 in the future if someone implements the necessary changes. In any of these cases, additional modifications to this driver would be required. However, by explicitly setting the version now, we can avoid potential issues and ensure compatibility going forward.

> Thus, I see no urgent need to do this, especially in scope of this PR.

While I understand your point that there's no urgent need to address this within the scope of this PR, I see many of the changes here as non-urgent — cleanups, small fixes, and code improvements. Adding this check aligns with the overall purpose of the PR.